### PR TITLE
[ENG 787] Fix CAISO file dating

### PR DESCRIPTION
**Pull request by Kladar**
## Summary
Updates the CAISO file-puller logic to check the latest valid value `Time` and checks it against the current CAISO time. If the latest value is later than the current time, it means the file is from yesterday, so update the datetime column accordingly.

Also adds some quality of life (which I have a bit of a habit of doing on PRs, hopefully nothing controversial).

## Relevant Issues (Optional)
Bug introduced in PR #431 

## Details
Quality of Life Stuff like 
- docstrings
- slight import renaming (import utils functions individually rather than as a module), 
- a tweak to ruff (since we are using black formatter for line-length, the ruff-linter is annoying right now). Recommend switching to ruff-format 

## Checklist

- [x] I have read and followed the [CONTRIBUTING.md](../CONTRIBUTING.md) guidelines
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added or updated necessary documentation
- [x] I have run the pre-commit hooks and addressed any issues
- [x] I have updated the CHANGELOG.md file if necessary

### CODE REVIEW

File: gridstatus/caiso.py
Comment (kmax12) [inline gridstatus/caiso.py:25]: why this change?

I don't feel strongly, but I like when I'm reading the code later in the file to be able to see that im using a repo-wide `util` function. 

also, this change broke the code: https://github.com/gridstatus/gridstatus/actions/runs/11113242953/job/30877128940?pr=442#step:7:31

File: pyproject.toml
Comment (kmax12) [inline pyproject.toml:85]: why add this to ignore?

Comment (Kladar) [reply pyproject.toml:85]: because we are using black, so black will try the formatting to the best of it's ability, but it falls down on comments and things

> Since you're using Black, you can just disable the Ruff E501 rule; Black will keep your lines under its line-length (as well as it can).

From [first answer here.](https://stackoverflow.com/questions/77396198/how-to-auto-format-e501-line-too-long-89-88-characters-in-pre-commit) 

Mostly I'd like to move to ruff-format (especially if we are using ruff-lint), but that feels like a separate PR. This is a  temporary quality of life don't-make-me-count-characters-even-if-black-can't-figure-out-a-docstring addition.

**Comment by Kladar** [comment]
I don't think it likes the hidden method `_handle_date` imported from utils. Fixing now.

File: gridstatus/caiso.py
Comment (WillKoehrsen) [inline gridstatus/caiso.py:1507]: Nit: `tz=self.default_timezone` to avoid using the string.

File: gridstatus/caiso.py
Comment (Kladar) [reply gridstatus/caiso.py:25]: yes saw that. Change was because the other imports were by function and that tends to be my preference, but will alter it back to avoid the code break.

Comment (Kladar) [reply gridstatus/caiso.py]: done.

File: pyproject.toml
Comment (WillKoehrsen) [reply pyproject.toml:85]: Should we move to only using `ruff`? It doesn't seem like you need both ruff and black.

File: gridstatus/caiso.py
Comment (Kladar) [reply gridstatus/caiso.py:1500]: makes sense, though `_get_historical()` is not a class method with a self param and is not currently being passed the timezone. 

see "timezone" params just below it doing a similar str reference. 

```
df["Time"] = df["Time"].apply(
        _make_timestamp,
        today=date,
        timezone="US/Pacific",
    )
```

Comment (Kladar) [reply gridstatus/caiso.py:1500]: Ah, should I do `CAISO.default_timezone`?

File: pyproject.toml
Comment (thehomebrewnerd) [reply pyproject.toml:85]: +1 to using only `ruff`. In a past life I've struggled with `ruff` and `black` wanting different things, so sticking with one tool seems good to me.

File: gridstatus/caiso.py
Comment (Kladar) [reply gridstatus/caiso.py]: Pushed, lmk what you think.

File: docs/pull_request_template.md
Comment (Kladar) [reply docs/pull_request_template.md:1]: Totally fair, I kind of the let copilot go ham on "make me a PR template" haha, can definitely pare it back. 

I kind of like Summary and Details for the reason I used it, like mention quality of life improvements in the summary and details them in details, but can definitely make it optional and lose the Resources and Checklist portions. I'll push another and you can let me know what you think.

File: gridstatus/utils.py
Comment (WillKoehrsen) [inline gridstatus/utils.py:352]: Can this be simplified to `df['Time'].max()`?

File: pyproject.toml
Comment (WillKoehrsen) [reply pyproject.toml:85]: > I can make that change in the next PR.

That would be great, thanks!

File: gridstatus/caiso.py
Comment (WillKoehrsen) [inline gridstatus/caiso.py:1472]: I don't have strong feelings about which to use, but I think we should standardize on function doc strings. [Here are a few options](https://stackoverflow.com/a/24385103/5755357) of which I slightly prefer Google's.

File: gridstatus/caiso.py
Comment (WillKoehrsen) [inline gridstatus/caiso.py:1503]: Do we need this logic for all datasets that use `_get_historical`, or only for the two

File: gridstatus/caiso.py
Comment (WillKoehrsen) [reply gridstatus/caiso.py]: > though _get_historical() is not a class method

My fault. Looks like you got it.

File: docs/pull_request_template.md
Comment (WillKoehrsen) [reply docs/pull_request_template.md:1]: The updated version is good. I like checklists provided all the items are actually necessary.

File: gridstatus/utils.py
Comment (Kladar) [reply gridstatus/utils.py:352]: I don't think so (that was my initial approach), but typically there is a timestamp for every time in the day because it also has the day ahead forecast for the value. The file updates the value column incrementally for what we care about. So we need the time at which we have the latest value. `max()` would return the 23:55 timestamp each time. 

Example 
CleanShot 2024-09-30 at 15 09 35@2x

I haven't checked this comprehensively for all methods that use `_get_historical()` though (`fuelsource`, `demand`, `storage`), so I will do that now. `max()` may work for some, but conceivably this implementation handles both cases.

File: gridstatus/caiso.py
Comment (Kladar) [reply gridstatus/caiso.py:1472]: Oop good catch, I also prefer google's (with types) and forgot to update my autodocstring extension. 

CleanShot 2024-09-30 at 15 24 15@2x

Highly recommend this though. https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring

**Comment by Kladar** [comment]
Also going to investigate these relevant tests a bit 
CleanShot 2024-09-30 at 15 39 37@2x

File: gridstatus/utils.py
Comment (WillKoehrsen) [reply gridstatus/utils.py:352]: Thanks for the image, that's helpful. I couldn't figure out what `value_col` was being passed from the caller because we call this with `utils.check_latest_value_time(df, file)` where evidently `file` is a `str`.

Comment (Kladar) [reply gridstatus/utils.py]: Okay, I feel this works now as a general solution for `storage`, `fuelsource`, and `demand`, which are the 3 files that take this data in. Hard to test outside of production workflows given the tendency to not work in the ~mornings and to know for sure, but methods are working.

CleanShot 2024-09-30 at 18 23 01@2x

@kmax12

Comment (Kladar) [reply gridstatus/utils.py]: @WillKoehrsen yeah also took me a bit to realize `file` is meaning the file from caiso sans the .csv, in the request URL. 

Also I was assuming that the file and the value_col were going to be the same, but for instance the `demand` file has the column `Current demand`, so I wrote a small mapping dictionary in that util method to get the correct column name depending on the file. That feels incredibly brittle, but at least gets us off performing the morning data surgery.

File: gridstatus/utils.py
Comment (kmax12) [inline gridstatus/utils.py:352]: i think it's more maintainable to add another parameter to `_get_historical` that is the column to use for this calculation.

we would then require all the callers of `_get_historical` (e.g `_get_historical_fuel_mix`, `_get_historical_load`) to provide the column they want to use. 

by doing this, if we were to add another endpoint like https://www.caiso.com/outlook/sp/history/20240929/rtm_forecast.csv, we wouldn't have to touch any of these lower level util functions

File: gridstatus/caiso.py
Comment (kmax12) [inline gridstatus/caiso.py:1503]: the tests are failing because the logic here is flawed. 

we only want to do this check if we are querying for the non-dated file.

File: gridstatus/utils.py
Comment (Kladar) [reply gridstatus/utils.py]: done

File: gridstatus/caiso.py
Comment (Kladar) [reply gridstatus/caiso.py]: added logic to check only if it's the non-dated.

File: pyproject.toml
Comment (Kladar) [reply pyproject.toml:85]: Tracked in ENG-797

**Comment by kmax12** [comment]
> 1 hr for checks is too long to be a useful iteration to test for code regressions in PRs.

a good release valve right now is to run the tests on the datasets your PR touches locally. that is pretty good for iterating quickly. 

getting tests running faster can be goal of future iteration. the tests take a long time because they depend on fetching data. any fast tests will have to bypass that as you suggest.

File: gridstatus/caiso.py
Comment (kmax12) [inline gridstatus/caiso.py:1475]: missing the docstring for column

**Review by kmax12 — APPROVED** [review]
things are looking good. given that all the caiso tests are passing locally, im good to merge this in without waiting for github CI. then, we can update our internals systems to use this version of `gridstatus`. thanks for fixing!

**Review by WillKoehrsen — APPROVED** [review]

File: gridstatus/utils.py
Comment (WillKoehrsen) [reply gridstatus/utils.py]: I see this was addressed by adding a `column` parameter to the function which makes sense.

**Comment by WillKoehrsen** [comment]
For future PRs, can we Squash and Merge?

**Comment by kmax12** [comment]
I updated repo settings to only allow squash merging now. dont know why it wasnt previously set up like that

image

**Comment by Kladar** [comment]
I will tend to delete my branches once they are merged, which can be an automatic setting as well.

**Pull request merged**
